blocks/graph.py

```diff
@@ -161,8 +162,40 @@ def replace(self, replacements):
             substitutes.
 
         """
-        return ComputationGraph(theano.clone(self.outputs,
-                                             replace=replacements))
+        # Due to theano specifics we have to make one replacement in time
+        replacements = OrderedDict(replacements)
+        # Replacements with previous replacements. We have to track
+        # variables in the new graph corresponding to original replacements.
```

Comment:
Thanks, this was helpful and now I understand a bit more. Will try to 
read your code again.

On 09.03.2015 16:11, dmitriy-serdyuk wrote:
>
> In blocks/graph.py 
> <https://github.com/bartvm/blocks/pull/407#discussion_r26043240>:
>
> > @@ -161,8 +162,40 @@ def replace(self, replacements):
> >              substitutes.
> >
> >          """
> > -        return ComputationGraph(theano.clone(self.outputs,
> > -                                             replace=replacements))
> > +        # Due to theano specifics we have to make one replacement in time
> > +        replacements = OrderedDict(replacements)
> > +        # Replacements with previous replacements. We have to track
> > +        # variables in the new graph corresponding to original replacements.
>
> I wanted to explain, that when I'm doing one replacement a time, I 
> need to do |replace| in both cost computation graph and all 
> expressions in future replacements (in keys and values of dict 
> |replacements|).
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/bartvm/blocks/pull/407/files#r26043240>.
>